### ISSUE TITLE
[rewrite] An Error thrown in oninit or view in a root component forces the component to be recreated

### ISSUE THREAD

**Issue by veloce** [lhorie/mithril.js#1311]
### Description:

Like said in the title, if an error is thrown during the `oninit` phase, before the `view()` completes, it will force the component to be recreated in each subsequent `render()`, thus keeping calling `oninit` and throwing error.

It is problematic because `oninit` can be used to make xhr calls, or opening web workers, or whatever. You can flood a server or raise memory issues if it happens together with a logic somewhere else in the app that calls `redraw` a lot of times (see jsbin).

Such errors that occurs in `oninit` phase may be hard to see for the dev (I mean it's easy to let one slip through), and mithril should have a safeguard for this imo.
### Steps to Reproduce:

This jsbin shows the behaviour when going from a view which has a setInterval that triggers a redraw to show a counter. Going from that view to the view with a buggy component makes the `oninit` hooks of the buggy component execute every second.

``` js
m.route(document.body, '/', {
  '/': {
    oninit() {
      this.counter = 0
      setInterval(() => {
        this.counter++
        m.redraw()
      }, 1000)
    },
    view() {
      return m('div', this.counter)
    }
  },
  '/buggycomp': {
    oninit() {
      console.log('buggy component oninit called')

      // an error thrown here or in view is actually the same
      throw new Error('makes my oninit hook fires on every render')
    },
    view() {
      return m('div', 'buggy comp never rendered')
    }
  }
})


setTimeout(() => {
  m.route.set('/buggycomp')
}, 3000)
```

https://jsbin.com/tisezo/1/edit?js,console,output
### Expected:

I'd expect here that the `oninit` fires only once, and the app crashes (or maybe return to the previous view)
### Actual:

`oninit` fires every 1s, keeping throwing error

**Comment by veloce** [comment]
I updated the jsbin, it was wrong

**Comment by lhorie** [comment]
I wouldn't expect oninit to fire only once since it hasn't actually completed running. For example, can we really say this component is initialized?

```
var foo = undefined
var MyComp = {
  oninit: function() {
    this.items = m.request({method: "GET", url: "/things", data: foo.bar.baz})
  },
  view: function() {return "hi"}
}
m.mount(document.body, MyComp)
```

To go back to your scenario, under what circumstances would it be acceptable for an application to throw an uncaught exception and still be in state that can be considered stable in a production environment?

**Comment by spacejack** [comment]
Here's a fiddle that demonstrates spammy behaviour due to a mal-formed view: https://jsfiddle.net/panxp6vz/2/

(Open your console to observe the error loop.)

I can't really justify any reason why it would ever go into production, other than poor development practices. That happens though.

It's also kind of inconvenient to debug the cause of the error. And I suppose you may catch some flak for spamming an API even when you're in dev.

I'm not sure why `oninit` is repeatedly called when the view throws an error; there is likely some reason behind it that I don't understand.

**Comment by pygy** [comment]
@tivac on gitter:
> mithril could just abort everything if an oninit ever throws, but I dunno that it's any better

Likewise for the `view`...

**Comment by veloce** [comment]
@spacejack example is much more significant than mine, and that's what I wanted to point out: `oninit` can fire a lot of times very quickly ; it is annoying in dev because you can food a server and it makes it hard to debug.
I'd also feel safer if I knew such error was forbidden by mithril in production.

@lhorie I'm not sure to understand your question, but I think it's better to have the app crash because of the faulty code, than having oninit spam a server (and don't work as well)

**Comment by pygy** [comment]
@lhorie you could poison the root (`root.threw = true`) if a view or a hook throws an error, and refuse to render afterwards.

**Comment by isiahmeadows** [comment]
@pygy @lhorie Wouldn't it be better to log it and just poison the *current* redraw attempt, instead of crashing the entire app?

**Comment by spacejack** [comment]
Ran into this today again. The debugging experience really isn't great, would love to have the re-create loop prevented if possible. I think a lot of newcomers to the library will run into this.

**Comment by lhorie** [comment]
Fixed in 8b168deef456b9b05aed2ef6f80411c733e578f2
